### PR TITLE
Personal Tokens

### PR DESCRIPTION
**Pull request by proofoftom**
##### Description

The merge of all things Gitcoin personal tokens.

##### Refers/Fixes

This adds the ability for one to create a personal token, and for other Gitcoin members to purchase and redeem those tokens for one's time.

##### Testing

Running QA document here: https://hackmd.io/K3DRZVwCSsuneF29wdCxvg?view

### CODE REVIEW

File: app/ptokens/models.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+            web3 = get_web3(self.network)
+            contract = web3.eth.contract(Web3.toChecksumAddress(self.token_address), abi=PTOKEN_ABI)
+            decimals = 10 ** contract.functions.decimals().call()
+            balance = contract.functions.balanceOf(holder_address).call() // decimals
+            self.cached_balances[str(holder_profile.id)] = balance
+            self.save()
+
+    def update_tx_status(self):
+        # Get transaction status
+        from dashboard.utils import get_tx_status
```
Comment (danlipert) [inline app/ptokens/models.py:228]: Can we move these imports?

**Review by danlipert — CHANGES_REQUESTED** [review]
Looking pretty good! Just a few comments/questions - also, lets not forget to add the `is_staff` templatetag around the UI so that only staff can see it when it gets rolled to production next week

File: app/app/urls.py [hunk 2]
```diff
@@ -79,6 +80,19 @@
     # board
     re_path(r'^dashboard/?', dashboard.views.board, name='dashboard'),
 
+    # personal_tokens
+    re_path(r'^ptoken/quickstart/?', ptokens.views.quickstart, name='ptoken_quickstart'),
+    re_path(r'^ptoken/faq/?', ptokens.views.faq, name='ptokens_faq'),
+    path('ptokens/redemptions/<int:redemptionId>/', ptokens.views.ptoken_redemption, name='token_redemption'),
+    path('ptokens/<int:tokenId>/purchase/', ptokens.views.ptoken_purchases, name='token_purchase'),
```
[the remark this answers is not in the document]
Comment (mds1) [reply app/app/urls.py]: Fixed in 34f55140b08c1f940fbde01f5754255cd44b4d70

File: app/ptokens/migrations/0008_auto_20200814_0035.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 2.2.4 on 2020-08-14 00:35
```
[the remark this answers is not in the document]
Comment (mds1) [reply app/ptokens/migrations/0008_auto_20200814_0035.py]: Resolved in 6dc76c8268305c27f1544287fa192c08c63c9b99

File: app/ptokens/models.py (hunk 1 above)
Comment (mds1) [reply app/ptokens/models.py]: Resolved in 34f55140b08c1f940fbde01f5754255cd44b4d70

File: app/dashboard/views.py [hunk 4]
```diff
@@ -912,6 +914,9 @@ def users_fetch_filters(profile_list, skills, bounties_completed, leaderboard_ra
         profile_list = profile_list.filter(
             hackathon_registration__hackathon=hackathon_id
         )
+    if only_with_tokens:
+        token_ids = PersonalToken.objects.filter(network=network).values_list('token_owner_profile_id', flat=True)
```
[the remark this answers is not in the document]
Comment (mds1) [reply app/dashboard/views.py]: Resolved in 34f55140b08c1f940fbde01f5754255cd44b4d70

File: app/dashboard/models.py [hunk 5]
```diff
@@ -4144,6 +4164,10 @@ def to_dict(self):
             dict: The profile card context.
 
         """
+        # Avoid circular ref error
+        from ptokens.models import PurchasePToken
```
[the remark this answers is not in the document]
Comment (mds1) [reply app/dashboard/models.py]: Resolved in 34f55140b08c1f940fbde01f5754255cd44b4d70

**Comment by mds1** [comment]
>  also, lets not forget to add the is_staff templatetag around the UI so that only staff can see it when it gets rolled to production next week

Hey @danlipert—all of your comments, including this quoted one, have been addressed by @zoek1 in commits 6dc76c8268305c27f1544287fa192c08c63c9b99 and 34f55140b08c1f940fbde01f5754255cd44b4d70.

You'll see there's also been a few new commits since your first review, so you may want to review those also

**Comment by thelostone-mc** [comment]
could you guys get the CI to pass as well

**Comment by mds1** [comment]
@thelostone-mc Sure thing. It looks it there's three places the 3 CI failed in. The ESLint one is an easy fix. The other two I'm less familiar with:

- `The command "pytest -p no:ethereum -p no:warnings" exited with 1.`
- `The command "pydocmd build" exited with 1.`

It seems those both failed in the same place in the codebase with `TypeError: Plain typing.NoReturn is not valid as type argument`, but it's not clear to me where exactly the problem is. Any insight into what needs to be done to resolve these, and how I can confirm locally before pushing the updates would be appreciated!

**Comment by thelostone-mc** [comment]
hmm i think those are related to 
Collecting ipfshttpclient in base.txt
Downloading ipfshttpclient-0.6.1-py3-none-any.whl (231 kB)

I've got the same thing on master! Trying to see if pinning it 0.6.0 fixes it

**Comment by mds1** [comment]
Interesting, let me know! We didn't change `base.txt` at all so it seems those may be unrelated to this PR?

File: app/assets/v2/js/user_card.js [hunk 6]
```diff
@@ -100,7 +107,15 @@ const renderPopOverData = function(data) {
...
   );
 
-  let mount_graph = [ tips_total_percent, bounties_total_percent, grants_total_percent ];
+  let mount_graph;
+
+  if (data.profile.has_ptoken) {
+    mount_graph = [ tips_total_percent, token_total_percent, bounties_total_percent, grants_total_percent ];
+  } else {
+    mount_graph = [ tips_total_percent, bounties_total_percent, grants_total_percent ];
+  }
```
Comment (octavioamu) [inline app/assets/v2/js/user_card.js:116]: maybe something more cleaner like 
`if (data.profile.has_ptoken) {
    mount_graph.push(token_total_percent)
}`

**Comment by thelostone-mc** [comment]
@mds1 I've fixed it up on master !
- Could you rebase it against master
- resolve conflicts

Also did we add the staff flags ?

**Review by octavioamu — CHANGES_REQUESTED** [review]
just left an small comment. Looking nice guys

File: app/townsquare/management/commands/payout_mini_clr.py [hunk 7]
```diff
@@ -62,7 +62,7 @@ def handle(self, *args, **options):
         network = 'mainnet' if not settings.DEBUG else 'rinkeby'
         from_address = settings.MINICLR_ADDRESS
         from_pk = settings.MINICLR_PRIVATE_KEY
-        DAI_ADDRESS = '0x6b175474e89094c44da98b954eedeac495271d0f' if network=='mainnet' else '0x6A9865aDE2B6207dAAC49f8bCba9705dEB0B0e6D'
+        DAI_ADDRESS = '0x6b175474e89094c44da98b954eedeac495271d0f' if network=='mainnet' else '0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa'
```
Comment (thelostone-mc) [inline app/townsquare/management/commands/payout_mini_clr.py:65]: ^ Was this update intended ?

File: app/app/fixtures/grants.json [hunk 8]
```diff
@@ -213,7 +213,7 @@
       "real_period_seconds": "2592000",
       "frequency_unit": "days",
       "frequency": "30",
-      "token_address": "0x6A9865aDE2B6207dAAC49f8bCba9705dEB0B0e6D",
+      "token_address": "0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa",
       "token_symbol": "DAI",
       "gas_price": "25000.0000",
       "new_approve_tx_id": "0x980ce738853f454c83d5551dc086a29d2d89781fb041a9ce581474ffdf639de4",
```
Comment (thelostone-mc) [inline app/app/fixtures/grants.json:219]: ^ could we undo all these changes

File: app/economy/models.py [hunk 9]
```diff
@@ -66,12 +66,12 @@ def default(self, obj):
...
+    return timezone.now()
 
 
 def get_0_time():
     """Get the local time."""
-    return localtime(timezone.datetime(1970, 1, 1).replace(tzinfo=pytz.utc))
+    return timezone.datetime(1970, 1, 1).replace(tzinfo=pytz.utc)
 
 
 class SuperModel(models.Model):
```
Comment (thelostone-mc) [inline app/economy/models.py:77]: ^ can these be undone ?

File: app/assets/v2/js/user_card.js (hunk 6 above)
Comment (mds1) [reply app/assets/v2/js/user_card.js:116]: I'll defer to @zoek1 on this one

File: app/townsquare/management/commands/payout_mini_clr.py (hunk 7 above)
Comment (mds1) [reply app/townsquare/management/commands/payout_mini_clr.py:65]: Yea this was intended.

The first address `0x6A9865aDE2B6207dAAC49f8bCba9705dEB0B0e6D` is the official Rinkeby DAI address. But Oasis on Rinkeby has been down for a while now, so it's very difficult to mint DAI on RInkeby.

We switched to `0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa`, which is the address for Rinkeby DAI given out when using the faucet available on https://app.compound.finance/. This makes it a lot easier to get Rinkeby DAI for development/testing

File: app/app/fixtures/grants.json (hunk 8 above)
Comment (mds1) [reply app/app/fixtures/grants.json:219]: See above comment https://github.com/gitcoinco/web/pull/7212#discussion_r483652209 — I think we should keep these changes.

File: app/economy/models.py (hunk 9 above)
Comment (mds1) [reply app/economy/models.py:77]: This change was by @zoek1, so I'll defer to him about if it should be undone or kept

Comment (zoek1) [reply app/economy/models.py]: done

File: app/assets/v2/js/user_card.js (hunk 6 above)
Comment (zoek1) [reply app/assets/v2/js/user_card.js]: done

**Comment by mds1** [comment]
Hey @thelostone-mc and @octavioamu — just add one more commit which updates the Terms of Service to add a new section based on what @owocki provided. Section 12 is the new section. Can you quickly review to confirm that all re-numberings of sections 13 through 30 were done properly? Just want to make sure there's none I missed. Thanks!

File: app/app/fixtures/grants.json (hunk 8 above)
Comment (thelostone-mc) [reply app/app/fixtures/grants.json:219]: gotcha

**Review by octavioamu — APPROVED** [review]

**Review by thelostone-mc — APPROVED** [review]

**Pull request merged**